Editor: add blogging prompts to posts

creativecoder opened the pull request:
#### Changes proposed in this Pull Request:

* Adds an experimental editor extension that displays a daily writing prompt for new posts.

| **Placeholder** | **Pullquote** |
| - | - |
| <img width="1053" alt="image" src="https://user-images.githubusercontent.com/1699996/196806921-250b1731-719a-4597-8659-214a480cf146.png"> | <img width="1054" alt="image" src="https://user-images.githubusercontent.com/1699996/196806942-c6606bfb-1512-4943-ac13-8bf0ad4fe4ac.png"> |

#### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion

pe7F0s-gI-p2

#### Does this pull request change what data or activity we track or use?

No.

#### Testing instructions:

**Jetpack Site**

- Be sure the site is connected to WP.com to load Jetpack blocks in the editor
- Enable beta blocks by adding `define( 'JETPACK_BETA_BLOCKS', true )` (or Settings > Jetpack Constants if you're using jurassic.ninja)
- Set the site as "having a blog" by doing at least one of the following
    - Set the `site_intent` option to `write`
    - Set the front page to show posts in Settings > Reading
    - Set a static page for the posts page in Settings > Reading
- Create a new post, you should see a writing prompt in the placeholder for the first paragraph
- Create a new post and add the `answer_prompt` query param with the prompt id for today as the value (e.g. `/wp-admin/post-new.php?answer_prompt=2129`), and you should see a writing prompt as a pullquote block, the tag `dailyprompt` added to the post, and the `_jetpack_blogging_prompt_key` post meta set to the prompt id.
- Create a new page or custom post type, and you should _not_ see the writing prompt
- Open an existing post, and you should _not_ see the writing prompt

**Simple Site**

- Sandbox a Simple site domain
- Apply this PR to your sandbox: `bin/jetpack-downloader test jetpack try/paragraph-block-writing-prompts`
- Continue with the testing instructions, as above.

creativecoder committed:
[not verified] Add writing prompts to posts

creativecoder committed:
[not verified] changelog

creativecoder committed:
[not verified] Removes logging

creativecoder committed:
[not verified] Removes uneeded index.js file

creativecoder committed:
[not verified] Use wp_rand for PHP < 7 compatibility

roo2 commented on projects/plugins/jetpack/extensions/blocks/writing-prompts/writing-prompts.php:
```diff
@@ -0,0 +1,68 @@
+const BLOCK_NAME   = 'jetpack/' . FEATURE_NAME;
+
+/**
+ * Registers Anchor.fm integration for the block editor.
```
anchor.fm stuff was left over from a copy I think

roo2 commented on projects/plugins/jetpack/extensions/blocks/writing-prompts/writing-prompts.php:
```diff
@@ -0,0 +1,68 @@
+}
+
+/**
+ * Checks URL params to determine the Anchor integration action to perform.
```
Anchor.fm in comment

roo2 commented on projects/plugins/jetpack/extensions/blocks/writing-prompts/writing-prompts.php:
```diff
@@ -0,0 +1,68 @@
+<?php
+/**
+ * Anchor.fm integration.
```
anchor.fm in comment

roo2 commented on projects/plugins/jetpack/extensions/blocks/writing-prompts/writing-prompts.php:
```diff
@@ -0,0 +1,68 @@
+/**
+ * Anchor.fm integration.
+ *
+ * @since 9.3.0
```
need to check if this number is correct, I remember there was something we had to do to get the CI tests to validate this 🤔

roo2 commented on projects/plugins/jetpack/extensions/blocks/writing-prompts/writing-prompts.php:
```diff
@@ -0,0 +1,68 @@
+		__(
+			'What’s the most time you’ve ever spent apart from your
+		favorite person? Tell us about it',
+			'jetpack'
```
hmm is the translation domain just 'jetpack' or should it be 'jetpack-writing-prompts'? I suppose because it's only a block and not a package it should just come under 'jetpack'?

creativecoder replied:
That's my understanding--this lives in the Jetpack plugin, so the original 'jetpack' text domain is used.

roo2 commented on projects/plugins/jetpack/extensions/blocks/writing-prompts/writing-prompts.php:
```diff
@@ -0,0 +1,68 @@
+
+	$random_prompt = $prompts[ wp_rand( 0, count( $prompts ) - 1 ) ];
+
+	wp_localize_script( 'jetpack-blocks-editor', 'Jetpack_WritingPrompts', array( 'prompt' => esc_html( $random_prompt ) ) );
```
hmmm I'm not sure if localizing `jetpack-blocks-editor` is correct, it is also used in anchor-fm, but the tilled-gallery block uses its own script called `jetpack-gallery-settings`  https://github.com/Automattic/jetpack/blob/trunk/projects/plugins/jetpack/extensions/blocks/tiled-gallery/tiled-gallery.php#L67

creativecoder replied:
In this case, I wanted to hook into the [aggregated editor script](https://github.com/Automattic/jetpack/blob/try/paragraph-block-writing-prompts/projects/plugins/jetpack/class.jetpack-gutenberg.php#L645) that Jetpack enqueues, which contains Jetpack blocks and editor extensions, including the writing prompts extension.

But I'm glad you pointed this out, since I think it's preferred to use `wp_add_inline_script` now for things that aren't actually localization--I'll refactor.

roo2 commented on projects/plugins/jetpack/extensions/blocks/writing-prompts/writing-prompts.php:
```diff
@@ -0,0 +1,68 @@
+		__( '“And they lived happily ever after.” Think about this line for a few minutes. Are you living happily ever after? If not, what will it take for you to get there?', 'jetpack' ),
+		__( 'Write about the last disagreement you had with a friend or family member — from their perspective.', 'jetpack' ),
+		__( 'You have the choice to erase one incident from your past, as though it never happened. What would you erase and why?', 'jetpack' ),
+		__( 'You’ve being exiled to a private island, and your captors will only supply you with five foods. What do you pick?', 'jetpack' ),
```
there's a grammar error in this one "`You've being exiled`"

roo2 commented on projects/plugins/jetpack/extensions/blocks/writing-prompts/writing-prompts.php:
```diff
@@ -0,0 +1,68 @@
+		favorite person? Tell us about it',
+			'jetpack'
+		),
+		__( 'What’s the most time you’ve ever spent apart from your favorite person? Tell us about it', 'jetpack' ),
```
duplicate prompt

roo2 commented on projects/plugins/jetpack/extensions/blocks/writing-prompts/writing-prompts.php:
```diff
@@ -0,0 +1,68 @@
+	if ( ! $post || ! $post->ID ) {
+		return;
+	}
+
```
We need to only show this for `new` posts as well, currently it's showing when you edit an existing post.
<img width="344" alt="Screen Shot 2022-10-13 at 3 10 47 pm" src="https://user-images.githubusercontent.com/22446385/195507131-be785baf-e612-4f34-8ca8-927eb25a753c.png">

creativecoder replied:
Thanks for catching! I had this check in and must have deleted it--will add it back.

roo2 reviewed (commented):
hmm it seems strange to me that this is implemented as an experimental block. Maybe a standalone package might make more sense? I'm not sure, I wonder what the jetpack crew will say.

creativecoder commented on projects/plugins/jetpack/extensions/index.json:
```diff
@@ -50,7 +50,7 @@
 		"videopress/video-chapters"
 	],
-	"experimental": [ "anchor-fm", "premium-content", "conversation", "dialogue" ],
+	"experimental": [ "anchor-fm", "premium-content", "conversation", "dialogue", "writing-prompts" ],
```
Note that including writing prompts as an experimental block will limit it to Simple and Atomic sites, and exclude it from self-hosted Jetpack (unless opted in using the `JETPACK_EXPERIMENTAL_BLOCKS` constant).

creativecoder replied:
> it seems strange to me that this is implemented as an experimental block

@roo2, see my comment above: this let's us create an editor extension that targets WP.com hosted sites. I think the editor extensions are intended to go in their own monorepo project, eventually, but still live within the main Jetpack plugin, for now.

roo2 replied:
ok, makes sense to me 👍

creativecoder committed:
[not verified] Addresses feedback

creativecoder committed:
[not verified] Retrieves writing prompts from the REST API

creativecoder committed:
[not verified] Limits writing prompts to posts only

roo2 commented on projects/plugins/jetpack/extensions/blocks/writing-prompts/editor.js:
```diff
@@ -0,0 +1,27 @@
+	const data = window.Jetpack_WritingPrompts;
+	const urlQuery = new URLSearchParams( document.location.search );
+	const embedPrompt = !! urlQuery.get( 'embed_prompt' );
+	if ( typeof data !== 'object' || ! data.prompts || data.prompts.length < 1 ) {
```
nit:

I just checked in my browser, if `data.prompts.length` were undefined, `data.prompts.length < 1` would return `false`, better to check for `data.prompts[ 0 ]` directly
```suggestion
	if ( typeof data !== 'object' || ! data.prompts || ! data.prompts[ 0 ] ) {
```

creativecoder replied:
Thanks for catching--updated.

roo2 commented on projects/plugins/jetpack/extensions/blocks/writing-prompts/writing-prompts.php:
```diff
@@ -0,0 +1,82 @@
+	$response = \Automattic\Jetpack\Connection\Client::wpcom_json_api_request_as_user(
+		'/sites/' . $blog_id . '/blogging-prompts?from=' . $today . '&number=1',
+		'v2',
+		array(),
```
In other places where we make api requests we set the X-Forwarded-For header e.g.
```
			array(
				'method'  => 'GET',
				'headers' => array(
					'Content-Type'    => 'application/json',
					'X-Forwarded-For' => ( new Visitor() )->get_ip( true ),
				),
			)
```

creativecoder replied:
Nice find--I was unaware of that convention. Now updated.

roo2 commented on projects/plugins/jetpack/extensions/blocks/writing-prompts/writing-prompts.php:
```diff
@@ -0,0 +1,82 @@
+		'v2',
+		array(),
+		null,
+		'wpcom'
```
it looks like `null`, and `'wpcom'` are the default arguments anyway.

creativecoder replied:
Fair point--for me being explicit here makes the code more readable, as there's no question what API namespace is being used.

roo2 reviewed (commented):
One thing I noticed, if I were to start a post, then save it as a draft and then come back to it, I'd no longer see the writing prompt. I don't think it's worth handling that case specially but just something I noticed. Appart from that I just had very minor nit picks.

creativecoder committed:
[not verified] Addresses feedback

roo2 commented on projects/plugins/jetpack/extensions/blocks/writing-prompts/writing-prompts.php:
```diff
@@ -0,0 +1,90 @@
+	if ( 'add' !== $current_screen->action || 'post' !== $current_screen->post_type ) {
+		return;
+	}
+
```
One option would be to limit this to only English language users which allows us to ship before we have translations in

roo2 commented on projects/plugins/jetpack/extensions/blocks/writing-prompts/editor.js:
```diff
@@ -0,0 +1,27 @@
+	await waitForEditor();
+
+	const { insertBlocks } = dispatch( 'core/block-editor' );
+	const writingPromptBlocks = embedPrompt
```
I tested with `?embed_prompt=1` but it didn't seem to embed the content properly
<img width="732" alt="Screen Shot 2022-10-25 at 11 15 26 am" src="https://user-images.githubusercontent.com/22446385/197659313-931e8874-6fb3-4219-af6c-dbe0857904cd.png">

roo2 reviewed (commented):
Another thing I noticed with this approach is that when I hit the _backspace_ key when the post is empty, the prompt is removed.

https://user-images.githubusercontent.com/22446385/197660024-90dabd5e-97f1-4e6a-8467-8732f0e46db4.mov

roo2 commented on projects/plugins/jetpack/extensions/blocks/writing-prompts/editor.js:
```diff
@@ -0,0 +1,27 @@
+
+	const { insertBlocks } = dispatch( 'core/block-editor' );
+	const writingPromptBlocks = embedPrompt
+		? [ createBlock( 'core/pullquote', { value: prompt.text } ), createBlock( 'core/paragraph' ) ]
```
in fact, it looks (to me) like [the app inserts](https://github.com/wordpress-mobile/WordPress-iOS/blob/trunk/WordPress/Classes/Extensions/Post+BloggingPrompts.swift) the `prompt.content` which includes markup for a pullquote. That is another option, I actually think it  probably makes more sense to always use `prompt.text` for consistency though, what do you think?

creativecoder replied:
I prefer telling the client to insert the relevant block(s).

I think sending block markup generated from one editor at a specific point in time in the past to another editor seems less robust then just telling the editor to insert a pullquote block with the specified content.

roo2 commented on projects/plugins/jetpack/extensions/blocks/writing-prompts/writing-prompts.php:
```diff
@@ -0,0 +1,90 @@
+	// Or if we aren't creating a new post.
+	if ( 'add' !== $current_screen->action || 'post' !== $current_screen->post_type ) {
+		return;
+	}
```
Oh also, we need to ensure the user was a blogger! maaybe with that `judge_potential_blogging_site` function you found, or will we need to duplicate the logic into jetpack? pe7F0s-mf-p2

creativecoder replied:
Yes, I've copied that function into Jetpack so we can use it here.

creativecoder committed:
[not verified] Rename feature to "Blogging" prompts for consistency with wpcom code

creativecoder committed:
[not verified] Only show writing prompt if site is potentially a blog

creativecoder committed:
[not verified] Adds compatibility for Simple Sites

creativecoder committed:
[not verified] Use "blogging" prompts everywhere, for consistency with existing code:wq

creativecoder commented:
Note that this PR should now be compatible with Simple Sites, as well. I've added testing instructions in the PR description.

There is a small change required to wpcom to get things working, which you can see in D91157-code. This change will need to be merged when this feature is released to wpcom.

roo2 reviewed (dismissed):
Tested on simple sites with D91157-code andmy jurassic.tube site with D88262-code. LGTM!

When we merge this we will be seeing the prompt, I think we need to make sure the translation issue is solved at least before we merge.

zinigor commented:
I have tested this one using my Jetpack development site, and I can't get it to work in the pullquote mode:
![изображение](https://user-images.githubusercontent.com/374293/199051075-1d1fde97-663b-4332-a7d6-a301a6b672ee.png)
The problem here is that the quote only contains the big `"` sign, and then it only says "Add a quote".

The other problem that I have experienced is that the quote that I get from the API is always the same:
`What does it mean to be a kid at heart?`

creativecoder committed:
[not verified] Adds meta and tag to posts that answer a blogging prompt

creativecoder committed:
[not verified] Allows specifying which checks are used to determine a site with a blog

creativecoder committed:
[not verified] Breaks out functions for determining a blogging site

creativecoder committed:
[not verified] Use _locale parameter for fetching prompts in site language

creativecoder committed:
[not verified] Show prompt by id

creativecoder committed:
Move blogging prompts extension to beta

creativecoder commented:
I've added a number of changes today:

Breaks out the blogging site checks so they can be used individually

- For example, the editor extension in this PR will load for `jetpack_has_write_intent` or `jetpack_has_posts_page`, but not `jetpack_has_or_will_publish_posts`, because we don't want to force the writing prompt just because a site has published posts, which could be used for things other than blogging.

Adds the `_locale` parameter to the blogging prompts api request, for compatibility with D91251-code and loading translated prompts (when they're ready).

Change the `answer_prompt` parameter to take a prompt id, so we can link the response to specific prompts, but also default to the first prompt if we can't find that id.

Moved the extension to `beta` rather than experimental, so we can merge this without having to wait for translations to be finished. We have other wpcom diffs that we want to use some of the blogging prompt functions, so I'd like to try and merge this soon-ish.

creativecoder commented:
@zinigor Thanks for testing!

> The problem here is that the quote only contains the big " sign, and then it only says "Add a quote".

I'm not able to replicate that, but I'll keep trying.

> The other problem that I have experienced is that the quote that I get from the API is always the same

That's expected: currently there is one prompt per day.

roo2 commented on projects/plugins/jetpack/_inc/blogging-prompts.php:
```diff
@@ -18,3 +18,180 @@ function jetpack_blogging_prompts_add_meta_data( $keys ) {
+ * }
+ * @return bool
+ */
+function jetpack_is_potential_blogging_site( $checks = array() ) {
```
It looks like we can just reuse the same logic as the mobile app p1667829421316519/1667217622.476049-slack-C03NLNTPZ2T

roo2 commented on projects/plugins/jetpack/_inc/blogging-prompts.php:
```diff
@@ -18,3 +20,180 @@ function jetpack_blogging_prompts_add_meta_data( $keys ) {
+
+	if ( jetpack_is_valid_blogging_prompt( $prompt_id ) ) {
+		update_post_meta( $post_id, '_jetpack_blogging_prompt_key', $prompt_id );
+		wp_add_post_tags( $post_id, 'dailyprompt' );
```
I think we also need to explicitly add to the `blogging_prompt_answers` table. Unless this already happens somewhere else? 
2dfa4-pb

creativecoder replied:
The answer should be recorded automatically using the `blogging-prompts` mu-plugin in wpcom (there are hooked functions that handle posts from both Simple and Atomic/Jetpack sites). The `_jetpack_blogging_prompt_key` meta is required for recording the answer post, so it can be tied to a specific prompt id.

roo2 replied:
Ahh thanks for explaining, yup I saw my post appear in 2e357-pb/ when I hit publish post 👍

roo2 commented on projects/plugins/jetpack/_inc/blogging-prompts.php:
```diff
@@ -18,3 +20,180 @@ function jetpack_blogging_prompts_add_meta_data( $keys ) {
+ * @return bool
+ */
+function jetpack_is_potential_blogging_site() {
+	return jetpack_has_write_intent() || jetpack_has_posts_page() || jetpack_has_or_will_publish_posts();
```
+1 for splitting it out like this even though we changed direction so it's not strictly necessary, I still think it makes it more readable. Related discussion was in this thread p1667869365431739/1667217622.476049-slack-C03NLNTPZ2T

roo2 commented on projects/plugins/jetpack/_inc/blogging-prompts.php:
```diff
@@ -18,3 +20,180 @@ function jetpack_blogging_prompts_add_meta_data( $keys ) {
+	);
+
+	if ( defined( 'IS_WPCOM' ) && IS_WPCOM ) {
+		// This will load the library, but the `enqueue_block_assets` hook is too late to load any endpoints
```
does this comment line still make sense? I suspect this might have been talking about a previous implementation?

creativecoder replied:
It's still correct, but I see how it's confusing now that the function is abstracted--I'll update the comment to try and make it clearer.

roo2 commented on projects/plugins/jetpack/_inc/blogging-prompts.php:
```diff
@@ -18,3 +20,180 @@ function jetpack_blogging_prompts_add_meta_data( $keys ) {
+		return true;
+	}
+
+	// They are choosing to set the posts to be set to 0.
```
I think this comment is incorrect, it is testing if the page_for_posts is NOT 0

creativecoder replied:
Thanks--copy/paste mistake--will correct.

roo2 commented on projects/plugins/jetpack/_inc/blogging-prompts.php:
```diff
@@ -18,3 +20,180 @@ function jetpack_blogging_prompts_add_meta_data( $keys ) {
+		return;
+	}
+
+	if ( jetpack_is_valid_blogging_prompt( $prompt_id ) ) {
```
This only checks if prompt_id matches todays prompt. I was about to say there would be an edge case if the user clicked a link from an old notification. But actually notifications are updated on page load so that they always show todays prompt. So this shouldn't be an issue :)

creativecoder replied:
Ideally we could fetch the prompt by id based on the value of `answer_prompt`, but the current API endpoint doesn't work that way. So for now, I'm just checking to make sure the prompt id is known, before recording an answer.

You bring up a good point about outdated links--I can have the API request look back one day to increase the possibility of finding the prompt id.

roo2 replied:
> but the current API endpoint doesn't work that way

Ok, makes sense

> I can have the API request look back one day to increase the possibility of finding the prompt id

Probably doesn't make sense to add the complexity though, I also think that would be a rare edge case

I think the ideal solution would be what you suggested fetching the prompt based on the ID from answer_prompt but if that doesn't exist yet, I don't think it's critical to add that. Possibly it could be a future improvement, I think it would make sense for the API to have that method.

roo2 commented on projects/plugins/jetpack/_inc/blogging-prompts.php:
```diff
@@ -18,3 +20,180 @@ function jetpack_blogging_prompts_add_meta_data( $keys ) {
+ */
+function jetpack_setup_blogging_prompt_response( $post_id ) {
+	// phpcs:ignore WordPress.Security.NonceVerification.Recommended
+	$prompt_id = isset( $_GET['answer_prompt'] ) && absint( $_GET['answer_prompt'] ) ? absint( $_GET['answer_prompt'] ) : false;
```
So prompt_id is only set if the user clicked through from a notification or the my-home card right? but if the user just openned the editor for `/posts`, I think they would still see the daily prompt but it would not be recorded.

creativecoder replied:
That's right. When showing the prompt as a placeholder, we don't really have any way to judge if the post created is an answer to the prompt, or not. In order to not record many unrelated answers, the post is only associated as a prompt answer when we can detect intent, using the `answer_prompt` query param.

roo2 replied:
ahh of course, that makes sense 👍

roo2 commented on projects/plugins/jetpack/extensions/blocks/blogging-prompts/blogging-prompts.php:
```diff
@@ -0,0 +1,68 @@
+	Blocks::jetpack_register_block( BLOCK_NAME );
+
+	// Load the blogging-prompts endpoint here on init so its route will be registered.
+	// We can use it with `WPCOM_API_Direct::do_request` to avoid a network request on Simple Sites.
```
nice work BTW figuring out how to avoid the request on wpcom 👌

roo2 commented on projects/plugins/jetpack/extensions/blocks/blogging-prompts/blogging-prompts.php:
```diff
@@ -0,0 +1,68 @@
+ * @return bool
+ */
+function should_load_blogging_prompts() {
+	return jetpack_has_write_intent() ||
```
We got the ✔️ from @davemart-in to use the full function `jetpack_is_potential_blogging_site` here, I thought that would make things simpler/easier. Now that you've implemented it though, maybe it's better as you've done it 🤷 it LGTM either way. p1667829421316519/1667217622.476049-slack-C03NLNTPZ2T

creativecoder replied:
Let's keep it as is. If you look at the headstart annotations that are used on themes with new sites, many of them create 2 or more posts--this effectively opts in all sites that are created, which doesn't seem like what we want.

roo2 replied:
good point!

roo2 commented:
todo: I added feedback on the code and am retesting it now

roo2 commented:
It looks like there's an issue with the prompt having html special characters in the calypso editor (wp-admin seems fine)
<img width="1479" alt="Screen Shot 2022-11-15 at 4 27 22 pm" src="https://user-images.githubusercontent.com/22446385/201843356-e82debe1-3420-465c-ad89-adc03e90c93f.png">

roo2 commented on projects/plugins/jetpack/_inc/blogging-prompts.php:
```diff
@@ -18,3 +20,180 @@ function jetpack_blogging_prompts_add_meta_data( $keys ) {
+	// phpcs:ignore WordPress.Security.NonceVerification.Recommended
+	$prompt_id = isset( $_GET['answer_prompt'] ) && absint( $_GET['answer_prompt'] ) ? absint( $_GET['answer_prompt'] ) : false;
+
+	if ( ! jetpack_is_new_post_screen() || ! $prompt_id ) {
```
In the calypso editor, this is exiting early and not setting metadata because `jetpack_is_new_post_screen` is only checking for the wp-admin screen.

creativecoder replied:
I'm not seeing that, it logs true for me (tested a Simple site), which makes sense because it is the correct screen within the wp-admin iframe.

However, the query param does need to be passed through to the iframed editor. PR for that at https://github.com/Automattic/wp-calypso/pull/70052

roo2 replied:
ahh yep, I think that was the missing piece, nice work

roo2 replied:
I tested and merged Automattic/wp-calypso#70052

creativecoder committed:
[not verified] Corrects page_for_posts comment

creativecoder committed:
[not verified] Clarifies comment about loading direct request library

creativecoder committed:
Allows answering yesterday's prompt

creativecoder committed:
Defaults editor placeholder to today's prompt

creativecoder commented:
> It looks like there's an issue with the prompt having html special characters in the calypso editor (wp-admin seems fine)

Thanks for calling this out @roo2 . I'm noticing this everywhere (Simple, Jetpack, wp-admin, Calypso), for the prompts that have html encodable characters, like apostrophes.

It really comes down to the API response for writing prompts, which is escaping all html in the prompt content (which seems too extreme, to me). The previous diff I proposed relating to decoupling the API from excerpt functions actually fixes this problem as well, so let's try to address it there: D91748-code

roo2 reviewed (dismissed):
LGTM! and I retested including the new diffs on calypso. I wasn't able to test on my jurassic.tube site, I had some issues getting that working again 😕. D91748-code would be necessary before we merge (and also D91157-code)

eoigal commented on projects/plugins/jetpack/_inc/blogging-prompts.php:
```diff
@@ -18,3 +20,186 @@ function jetpack_blogging_prompts_add_meta_data( $keys ) {
+ *
+ * @return stdClass[] Array of blogging prompt objects.
+ */
+function jetpack_get_daily_blogging_prompts() {
```
I think we should allow a timestamp to be passed into this function. When rendering notifications, we will need to get the prompt to match the note timestamp similar to `\A8C\Writing_Prompts\Notification::get_blogging_prompt_by_note_timestamp`. 

I can update `\A8C\Writing_Prompts\Notification::get_blogging_prompt_by_note_timestamp` to use `jetpack_get_daily_blogging_prompts` if that arg is supported.

creativecoder replied:
I've modified so you can specify a unix timestamp--let me know if that will work!

davemart-in reviewed (dismissed):
Tested the following:

✅ - Add new post (showed as placeholder)
✅ - Add new page (did not show)
✅ - Tested answer_prompt (showed as pull quote, dailyprompt tag)
✅ - Existing page (no prompt was shown)
✅ - Published post containing pull quote

Everything looks good!

creativecoder committed:
Ensure mag16 language for blogging prompts

creativecoder committed:
Allow specifying the timestamp to get blogging prompts for that day

creativecoder commented:
Thanks for the team reviews! I think this is ready for a look by Jetpack Crew.

bindlegirl reviewed (dismissed):
Works as advertised.

creativecoder commented:
Note that there's a companion PR at https://github.com/Automattic/jetpack/pull/27483 to move this editor extension to experimental. We'd like to make it available to WP.com hosted sites (Simple and Atomic) for next week's release.

anomiex commented on projects/plugins/jetpack/_inc/blogging-prompts.php:
```diff
@@ -18,3 +20,212 @@ function jetpack_blogging_prompts_add_meta_data( $keys ) {
+ * @return void
+ */
+function jetpack_setup_blogging_prompt_response( $post_id ) {
+	// phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
Why is nonce verification being ignored here? We try to include a comment explaining it.

See p9dueE-4z8-p2 for more info.

creativecoder replied:
We are generating the link to be clicked from a WP.com notification, so it's possible the nonce would change before the user clicks the link. I'll add a comment.

Not ideal, as we are "setting a field," in the form of a tag, post_meta with the prompt id, and a prefilled pull quote block, though no data is saved until the user publishes. Please let me know if you have other thoughts or a better solution!

anomiex replied:
As long as it's well-explained in the comment, I think that's a good start. :+1:

Do I have this right? There's a link in the notification, and clicking the link winds up calling this function. You can't put a nonce in the link, because it might be a while in between when the link is generated and when it's clicked. The "attack" that a nonce would prevent would be some third party tricking the blog admin's browser into following the link, maybe repeatedly to flood the site with these posts. Ideally the comment should address why that potential attack isn't something to actually worry about.

anomiex commented on projects/plugins/jetpack/_inc/blogging-prompts.php:
```diff
@@ -18,3 +20,212 @@ function jetpack_blogging_prompts_add_meta_data( $keys ) {
+ */
+function jetpack_get_daily_blogging_prompts( $time = 0 ) {
+	// Default to the time in the site's timezone.
+	$timestamp   = $time ? $time : current_time( 'timestamp' ); // phpcs:ignore WordPress.DateTime.CurrentTimeTimestamp.Requested
```
Why are you ignoring this sniff? Wouldn't it be better to use `time()` and `wp_date()` instead?

As it is, you have `$time` documented as a Unix timestamp on line 57, but then on line 64 you're treating it as a deprecated "WordPress timestamp+offset" value.

creativecoder replied:
I didn't realize that wp_date will default to the site's timezone (which is what's desired here, to respect the site settings). Will refactor.

anomiex commented on projects/plugins/jetpack/_inc/blogging-prompts.php:
```diff
@@ -18,3 +20,212 @@ function jetpack_blogging_prompts_add_meta_data( $keys ) {
+	}
+
+	$blog_id = \Jetpack_Options::get_option( 'id' );
+	$path    = '/sites/' . $blog_id . '/blogging-prompts?from=' . $yesterday . '&number=10&_locale=' . $locale;
```
While all this is probably safe, it might be better to wrap the variables with `rawurlencode()` to guarantee it.

anomiex commented on projects/plugins/jetpack/_inc/blogging-prompts.php:
```diff
@@ -18,3 +20,212 @@ function jetpack_blogging_prompts_add_meta_data( $keys ) {
+		return null;
+	}
+
+	$body    = json_decode( wp_remote_retrieve_body( $response ) );
```
Should there be error checking here in case the `json_decode()` returns null?

anomiex commented on projects/plugins/jetpack/_inc/blogging-prompts.php:
```diff
@@ -18,3 +20,212 @@ function jetpack_blogging_prompts_add_meta_data( $keys ) {
+ *
+ * @return string
+ */
+function jetpack_get_mag16_locale() {
```
This function seems a bit unfortunate to have in Jetpack rather than having the server side do the fix-up. If you were to add fr_BE, for example, you'd have to wait for the monthly Jetpack release (and for people to update to it) for standalone sites to get it. While if this were done on the server side instead, they'd start getting it as soon as you updated it there.

creativecoder replied:
That's a good point. The blogging-prompts endpoint is relying on a global `_locale`, but I'll see about customizing it just for this endpoint and pulling out this function. FYI, some related conversation: D91251-code#1932120

anomiex commented on projects/plugins/jetpack/extensions/blocks/blogging-prompts/blogging-prompts.php:
```diff
@@ -0,0 +1,68 @@
+	if ( should_load_blogging_prompts() ) {
+		$daily_prompts = wp_json_encode( jetpack_get_daily_blogging_prompts() );
+
+		// See p7H4VZ-2cf-p2 for why prompt data is escaped this way.
```
As the post itself notes, `json_encode()` is safe to use here with the right flags, and the problem back in February 2019 was that WordPress still supported PHP 5.2 which didn't have those flags. We've been at a minimum of 5.6 now for a long time, so just use `wp_json_encode( ..., JSON_HEX_TAG )`.

creativecoder replied:
Thanks for clarifying what's currently considered safe. I'll refactor.

creativecoder committed:
Blogging prompts: move to experimental blocks (#27483)

* Blogging Prompts: Move to experimental blocks
* Simplifies date generation

anomiex commented on projects/plugins/jetpack/extensions/blocks/blogging-prompts/blogging-prompts.php:
```diff
@@ -0,0 +1,68 @@
+function should_load_blogging_prompts() {
+	return jetpack_has_write_intent() ||
+			jetpack_has_posts_page() ||
+			// phpcs:ignore WordPress.Security.NonceVerification.Recommended
```
As above, why can we ignore nonce verification here?

anomiex commented:
Hmm, my review came in a few minutes too late. :( Hopefully those comments get addressed in a followup.

creativecoder commented:
@anomiex Thanks so much for the review and feedback! I'll answer comments and create a follow-up PR to address issues later today.

creativecoder commented:
Follow-up PR here: https://github.com/Automattic/jetpack/pull/27557
